File: AStar/Python/node.py
from enum import Enum

# The Main class for the node which is the building block of the board.
class Node:
  """This constructor ensures that the node always has a valid position and type.

  :param type: The type of the node.
  :param x The X position of the node.
  :param y The Y position of the node.
  """
  def __init__(self, type, x, y):
    self.y = y
    self.x = x
    self.type = type

    self.cost = 0 # The cost of the node.
    self.g = 0 # The distance between the node and the starting node.
    self.h = 0 # The heuristic distance between the node and the destination node.
    self.f = 0 # The total cost of a node.
    self.parent = None # Parent node, used for backtracking the path.
    self.neighbours = list() # The neighbours of the node.

  def __str__(self):
      return self.type.value

class Types(Enum):
  FLOOR = "-"
  WALL = "X"
  START = "S"
  DESTINATION = "D"
  PATH = "*"

File: AStar/Ruby/node.rb
##
# The Main class for the node which is the building block of the board.

class Node

  ##
  # These variables can only be read outside of this class.
  # +x+:: The X position of the node.
  # +y+:: The Y position of the node.
  # +cost+:: The cost of the node.
  attr_reader :x, :y, :cost

  ##
  # These variables can be both read and changed outside of this class.
  # +type+:: The type of the node.
  # +parent:: Parent node, used for backtracking the path.
  # +neighbours+:: The neighbours of the node.
  # +g+:: The distance between the node and the starting node.
  # +h+:: The heuristic distance between the node and the destination node.
  # +f+:: The total cost of a node
  attr_accessor :type, :parent, :neighbours, :g, :h, :f

  ##
  # This constructor ensures that the node always has a valid position and type.
  # Params:
  # +type+:: The type of the node.
  # +x+:: The X position of the node.
  # +y+:: The Y position of the node.
  def initialize(type, x, y)
    @x = x
    @y = y
    @type = type

    @cost = 0
    @g = 0
    @h = 0
    @f = 0
  end

  ##
  # Returns the string representation of a node which depends on it's type.
  def to_s
    case @type
    when :wall
      'X'
    when :start
      'S'
    when :destination
      'D'
    when :path
      '*'
    when :checked
      '?'
    else
      '-'
    end
  end

end
